source/Higgs/UntruncatedAtMostTwo.lagda

```diff
@@ -0,0 +1,244 @@
+                x ≠ x'
+  repeat-distinct f =
+    let (x , x' , pf , _) = f-ph f
+    in pf
+
+  f₁ = constantly y
+  ix₁ = repeat-indices f₁
+  f₂ = almost-constantly X-discrete (pr₁ ix₁) y' y
+  ix₂ = repeat-indices f₂
+
+  I : y ＝ y' → ix₁ ＝ ix₂
+  I e = ap repeat-indices
+    (almost-constantly-is-constant X-discrete (pr₁ ix₁) y y' e)
+
+  II : (x : X)
+     → (pr₁ ix₁ ≠ x)
+     → y ＝ f₂ x
+  II x ne with X-discrete x (pr₁ ix₁)
+  ... | inl e = 𝟘-elim (ne (e ⁻¹))
+  ... | inr _ = refl
+
+  III : f₂ (pr₁ ix₁) ＝ y'
+  III with X-discrete (pr₁ ix₁) (pr₁ ix₁)
+  ... | inl e = refl
+  ... | inr ne = 𝟘-elim (ne refl)
+
+  IV : ix₁ ＝ ix₂ → y ＝ y'
+  IV e =
+    y            ＝⟨ II (pr₂ ix₁) (repeat-distinct f₁) ⟩
+    f₂ (pr₂ ix₁) ＝⟨ ap (f₂ ∘ pr₂) e ⟩
+    f₂ (pr₂ ix₂) ＝⟨ refl ⟩
+    f₂ (pr₂ (repeat-indices f₂)) ＝⟨ repeat-is-repeat f₂ ⁻¹ ⟩
+    f₂ (pr₁ (repeat-indices f₂)) ＝⟨ refl ⟩
+    f₂ (pr₁ ix₂) ＝⟨ ap (f₂ ∘ pr₁) (e ⁻¹) ⟩
+    f₂ (pr₁ ix₁) ＝⟨ III ⟩
+    y' ∎
+
+  V : is-decidable (ix₁ ＝ ix₂) → is-decidable (y ＝ y')
+  V (inl e) = inl (IV e)
+  V (inr ne) = inr (ne ∘ I)
+
+  VI : is-decidable (ix₁ ＝ ix₂)
+  VI = ×-is-discrete X-discrete X-discrete ix₁ ix₂
+
+\end{code}
+
+We may write the untruncated form of the at-most-2 lemma in this form
+
+\begin{code} 
+
+at-most-two-is-pigeonhole
+  : {𝓤 : Universe}
+  → {X : 𝓤 ̇}
+  → ((x y z : X) → (z ＝ x) + (x ＝ y) + (y ＝ z))
+  → (f : Fin 3 → X)
+  → f has-a-repetition
+at-most-two-is-pigeonhole at-most-2 f = II I
+ where
+  v1 v2 v3 : Fin 3
+  v1 = inr ⋆
+  v2 = inl (inr ⋆)
+  v3 = inl (inl (inr ⋆))
+
+  true-when : Fin 3
+            → Fin 3
+            → 𝓤 ⁺ ̇
+  true-when (inr _) (inr _) = 𝟙
+  true-when (inl (inr _)) (inl (inr _)) = 𝟙
+  true-when (inl (inl (inr _))) (inl (inl (inr _))) = 𝟙
+  {-# CATCHALL #-}
+  true-when _ _ = 𝟘
+
+  v3-not-1 : v3 ≠ v1
+  v3-not-1 e = 𝟘-elim (transport (true-when v3) e ⋆)
+  v1-not-2 : v1 ≠ v2
+  v1-not-2 e = 𝟘-elim (transport (true-when v1) e ⋆)
+  v2-not-3 : v2 ≠ v3
+  v2-not-3 e = 𝟘-elim (transport (true-when v2) e ⋆)
+
+  I : (f v3 ＝ f v1) + (f v1 ＝ f v2) + (f v2 ＝ f v3)
+  I = at-most-2 (f v1) (f v2) (f v3)
+
+  II : (f v3 ＝ f v1) + (f v1 ＝ f v2) + (f v2 ＝ f v3)
+     → f has-a-repetition
+  II (inl e31)       = ( v3 , v1 , v3-not-1 , e31 )
+  II (inr (inl e12)) = ( v1 , v2 , v1-not-2 , e12 )
+  II (inr (inr e23)) = ( v2 , v3 , v2-not-3 , e23 )
+
+aut-Ω-discrete-has-em
+  : is-discrete (Aut Ω)
+  → (𝕗 : Aut Ω)
+  → (𝕗 ＝ 𝕚𝕕) + (𝕗 ≠ 𝕚𝕕)
+aut-Ω-discrete-has-em aut-disc 𝕗 =
+  aut-disc 𝕗 𝕚𝕕
+
+untruncated-at-most-two-iff-em
+ : ((f g h : Aut Ω) → (h ＝ f) + (f ＝ g) + (g ＝ h))
+ ↔ ((𝕗 : Aut Ω) → (𝕗 ＝ 𝕚𝕕) + (𝕗 ≠ 𝕚𝕕))
+untruncated-at-most-two-iff-em = (FW , BW)
+ where
+  FW : ((f g h : Aut Ω) → (h ＝ f) + (f ＝ g) + (g ＝ h))
+     → ((𝕗 : Aut Ω) → (𝕗 ＝ 𝕚𝕕) + (𝕗 ≠ 𝕚𝕕))
+  FW at-most-two = aut-Ω-discrete-has-em
+    (at-most-discrete-gives-discrete
+      (Fin 3) (Aut Ω)
+      Fin-is-discrete
+      (at-most-two-is-pigeonhole at-most-two)
+    )
```

Comment:
Please close the bracket in the previous line.

Reply:
Done